### PR TITLE
Allow primitive array creation from iterators of PrimitiveTypes (as well as `Option`)

### LINKED ISSUE #1298: More ergonomic / idiomatic primitive array creation from iterators
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I want to write this to create an array without nulls (as I am used to using this syntax all over rust codebases): 

```rust
let value_data: Int8Array = [10_i8, 11, 12, 13, 14, 15, 16, 17]
  .into_iter()
  .collect();
```

However, i have to write this (add a `Some` so it becomes an `Option`) which is annoying.

```rust
let value_data: Int8Array = [10_i8, 11, 12, 13, 14, 15, 16, 17]
  .into_iter()
  .map(Some)
  .collect();
```

Note you can write this, but that is not as idomatic
```rust
        let value_data = Int8Array::from_iter_values([10_i8, 11, 12, 13, 14, 15, 16, 17]);
```

**Describe the solution you'd like**
But it would be more ergonomic to be able to just use the collect syntax

Namely I want  to write this and have it work:

```rust
let value_data: Int8Array = [10_i8, 11, 12, 13, 14, 15, 16, 17]
  .into_iter()
  .collect();
```

Also, for BooleanArray and StringArray

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1298.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
cc @alamb

**Comment by alamb** [comment]
Thanks @viirya -- Sorry for the delay in review; I was out last week so I am am still catching up -- I'll add this one to my review queue and hope to get to it shortly.

**Comment by viirya** [comment]
Thank you @alamb !

File: arrow/src/array/array_primitive.rs [hunk 1]
```diff
@@ -329,8 +330,53 @@ impl<'a, T: ArrowPrimitiveType> PrimitiveArray<T> {
     }
 }
 
-impl<T: ArrowPrimitiveType, Ptr: Borrow<Option<<T as ArrowPrimitiveType>::Native>>>
-    FromIterator<Ptr> for PrimitiveArray<T>
+#[derive(Debug)]
+pub struct ArrowPrimitiveTypeNative<T: ArrowPrimitiveType> {
```
Comment (alamb) [inline arrow/src/array/array_primitive.rs:334]:
```suggestion
struct ArrowPrimitiveTypeNative<T: ArrowPrimitiveType> {
```

I wonder if this needs to be `pub`? 

If it does need to be pub I think we should probably document it.

**Review by alamb — COMMENTED** [review]
This is very clever @viirya  -- thank you very much.

I have a concern (but no specific suggestions just yet)

Today if people `impl Borrow<i8> for MyAwesomeType` then I think (untested) they can create arrays from an iterator of `MyAwesomeType` which they might not be able to do after this PR 🤔  Though perhaps they just need to implement `ArrowPrimitiveTypeNative`

I tried a little to figure out a clever way to implement the conversion using a trait that is a superset of `Borrow` to make it backwards compatible and more general. Thus instead of

```rust
impl<T: ArrowPrimitiveType, Ptr: Borrow<Option<<T as ArrowPrimitiveType>::Native>>>
    FromIterator<Ptr> for PrimitiveArray<T>
```

It would look something like

```rust
impl<T: ArrowPrimitiveType, Ptr: BorrowAsOption<<T as ArrowPrimitiveType>::Native>>
    FromIterator<Ptr> for PrimitiveArray<T>
..

pub trait BorrowAsOption<T: ArrowPrimitiveType> {
    fn as_native(&self) -> Option<&<T as ArrowPrimitiveType>::Native>;
}
```

But my mind was being blown by type errors.

File: arrow/src/array/array_primitive.rs (hunk 1 above)
Comment (viirya) [reply arrow/src/array/array_primitive.rs:334]: If people want to use their types to create arrays from an iterator, they need to implement `From` for `ArrowPrimitiveTypeNative`. So I guess we need it to be `pub`? I will document it.

**Comment by viirya** [comment]
> Today if people `impl Borrow<i8> for MyAwesomeType` then I think (untested) they can create arrays from an iterator of `MyAwesomeType` which they might not be able to do after this PR 🤔 Though perhaps they just need to implement `ArrowPrimitiveTypeNative`

I guess you mean `impl Borrow<Option<i8>> for MyAwesomeType`? Yes, this is a concern. People need to implement `impl From<MyAwesomeType> for ArrowPrimitiveTypeNative<Int8Type>`.


> I tried a little to figure out a clever way to implement the conversion using a trait that is a superset of `Borrow` to make it backwards compatible and more general. Thus instead of
> 
> ```rust
> impl<T: ArrowPrimitiveType, Ptr: Borrow<Option<<T as ArrowPrimitiveType>::Native>>>
>     FromIterator<Ptr> for PrimitiveArray<T>
> ```
> 
> It would look something like
> 
> ```rust
> impl<T: ArrowPrimitiveType, Ptr: BorrowAsOption<<T as ArrowPrimitiveType>::Native>>
>     FromIterator<Ptr> for PrimitiveArray<T>
> ..
> 
> pub trait BorrowAsOption<T: ArrowPrimitiveType> {
>     fn as_native(&self) -> Option<&<T as ArrowPrimitiveType>::Native>;
> }
> ```
> 
> But my mind was being blown by type errors.

Not sure if I understand above way. Seems people still need to implement `BorrowAsOption`, so it's still not backwards compatible.

**Review by alamb — APPROVED** [review]
> Not sure if I understand above way. Seems people still need to implement BorrowAsOption, so it's still not backwards compatible.


Yeah, after more thought here I can't come up with anything better 

Some documentation comments on `ArrowPrimitiveTypeNative` would definitely be useful. I wonder if changing its name to something like `NativeAdapter` would make its purpose clearer.

I wonder if we need something similar for `str` and `bool` to make `StringArrays` and `BooleanArrays` (can do as a follow on PR)

maybe @shepmaster  or @carols10cents  has some alternate suggestions

**Comment by viirya** [comment]
Thanks @alamb ! Just renamed `ArrowPrimitiveTypeNative` to `NativeAdapter` and added some documentation.

**Comment by alamb** [comment]
Thanks @viirya  -- this looks great

**Comment by viirya** [comment]
Thanks @alamb for your review and suggestion!

**Pull request merged**